File: experiments/2d_visualization/test_lbg_triple_ring.py
"""Test LBG (Linde-Buzo-Gray) on triple ring data with GIF visualization."""

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
from PIL import Image

sys.path.insert(0, str(Path(__file__).parents[2] / "algorithms" / "lbg" / "python"))
sys.path.insert(0, str(Path(__file__).parents[2] / "data" / "2d"))

from model import LindeBuzoGray, LBGParams
from sampler import sample_triple_ring


# Triple ring geometry (matching C++ implementation)
TRIPLE_RING_PARAMS = [
    (0.50, 0.23, 0.06, 0.14),  # top center
    (0.27, 0.68, 0.06, 0.14),  # bottom left
    (0.73, 0.68, 0.06, 0.14),  # bottom right
]


def draw_triple_ring_background(ax) -> None:
    """Draw triple ring background using geometric shapes."""
    theta = np.linspace(0, 2 * np.pi, 100)
    for cx, cy, r_inner, r_outer in TRIPLE_RING_PARAMS:
        outer_x = cx + r_outer * np.cos(theta)
        outer_y = cy + r_outer * np.sin(theta)
        inner_x = cx + r_inner * np.cos(theta)
        inner_y = cy + r_inner * np.sin(theta)
        ax.fill(outer_x, outer_y, color="lightblue", alpha=0.3)
        ax.fill(inner_x, inner_y, color="white")


def create_frame(
    ax,
    points: np.ndarray,
    nodes: np.ndarray,
    edges: list[tuple[int, int]],
    iteration: int,
    extra_info: str = "",
) -> None:
    """Create a single frame for visualization."""
    ax.clear()

    # Draw geometric background
    draw_triple_ring_background(ax)

    ax.scatter(points[:, 0], points[:, 1], c="skyblue", s=3, alpha=0.3, label="Data")

    # LBG has no edges, just plot nodes
    ax.scatter(nodes[:, 0], nodes[:, 1], c="red", s=50, zorder=5, label="Nodes")

    ax.set_xlim(0, 1)
    ax.set_ylim(1, 0)
    ax.set_aspect("equal")

    title = f"LBG - Epoch {iteration} ({len(nodes)} nodes)"
    if extra_info:
        title += f" [{extra_info}]"
    ax.set_title(title)
    ax.legend(loc="upper right")


def run_experiment(
    n_samples: int = 1500,
    n_iterations: int = 100,  # LBG uses epochs, not individual iterations
    gif_frames: int = 50,
    output_gif: str = "lbg_triple_ring_growth.gif",
    output_final: str = "lbg_triple_ring_final.png",
    seed: int = 42,
) -> None:
    """Run LBG experiment with visualization."""
    np.random.seed(seed)

    # Sample points using mathematical formula
    print(f"Sampling {n_samples} points from triple ring (mathematical)...")
    points = sample_triple_ring(n_samples=n_samples, seed=seed)
    print(f"Sampled {len(points)} points")

    # LBG parameters
    params = LBGParams(
        n_nodes=50,
        max_epochs=n_iterations,
        convergence_threshold=1e-6,
        use_utility=False,
    )
    model = LindeBuzoGray(n_dim=2, params=params, seed=seed)

    frames = []
    frame_interval = max(1, n_iterations // gif_frames)

    fig, ax = plt.subplots(figsize=(8, 8), facecolor="white")

    def callback(m, epoch):
        if epoch % frame_interval == 0 or epoch == n_iterations - 1:
            nodes, edges = m.get_graph()
            create_frame(ax, points, nodes, edges, epoch)
            fig.canvas.draw()

            img = Image.frombuffer(
                "RGBA", fig.canvas.get_width_height(), fig.canvas.buffer_rgba()
            )
            frames.append(img.convert("RGB"))
            print(f"Epoch {epoch}: {len(nodes)} nodes")

    print(f"Training LBG for {n_iterations} epochs...")
    print(f"Parameters: n_nodes={params.n_nodes}")
    model.train(points, n_iterations=n_iterations, callback=callback)

    nodes, edges = model.get_graph()
    create_frame(ax, points, nodes, edges, model.n_learning)
    plt.savefig(output_final, dpi=150, bbox_inches="tight", facecolor="white")
    print(f"Saved final result: {output_final}")

    if frames:
        frames.extend([frames[-1]] * 10)
        frames[0].save(
            output_gif,
            save_all=True,
            append_images=frames[1:],
            duration=200,  # Slower for batch learning
            loop=0,
        )
        print(f"Saved GIF: {output_gif}")

    plt.close(fig)
    print("Done!")


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Test LBG on triple ring data")
    parser.add_argument("-n", "--n-samples", type=int, default=1500)
    parser.add_argument("--iterations", type=int, default=100)  # Epochs for LBG
    parser.add_argument("--frames", type=int, default=50)
    parser.add_argument("--output-gif", type=str, default="triple_ring_growth.gif")
    parser.add_argument("--output-final", type=str, default="triple_ring_final.png")
    parser.add_argument("--seed", type=int, default=42)

    args = parser.parse_args()

    run_experiment(
        n_samples=args.n_samples,
        n_iterations=args.iterations,
        gif_frames=args.frames,
        output_gif=args.output_gif,
        output_final=args.output_final,
        seed=args.seed,
    )

File: experiments/2d_visualization/cpp/test_lbg_triple_ring.cpp
/**
 * @file test_lbg_triple_ring.cpp
 * @brief Test LBG on triple ring data (C++ implementation)
 *
 * Generates triple ring sample data, runs LBG training,
 * and outputs results to CSV files for visualization.
 */

#include <chrono>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <vector>

#include <Eigen/Core>
#include "lbg.hpp"

namespace fs = std::filesystem;

/**
 * @brief Sample points from triple ring (3 separate rings in triangle pattern).
 */
std::vector<Eigen::Vector2f> sample_triple_ring(
    int n_samples,
    unsigned int seed)
{
    std::mt19937 rng(seed);
    std::uniform_real_distribution<float> angle_dist(0.0f, 2.0f * M_PI);

    // Three separate rings arranged in triangle pattern
    std::vector<std::tuple<float, float, float, float>> rings = {
        {0.50f, 0.23f, 0.06f, 0.14f},  // top center
        {0.27f, 0.68f, 0.06f, 0.14f},  // bottom left
        {0.73f, 0.68f, 0.06f, 0.14f},  // bottom right
    };

    std::vector<Eigen::Vector2f> samples;
    samples.reserve(n_samples);

    int samples_per_ring = n_samples / 3;

    for (const auto& [cx, cy, r_inner, r_outer] : rings) {
        std::uniform_real_distribution<float> radius_dist(r_inner, r_outer);
        for (int i = 0; i < samples_per_ring; ++i) {
            float angle = angle_dist(rng);
            float radius = radius_dist(rng);
            float x = cx + radius * std::cos(angle);
            float y = cy + radius * std::sin(angle);
            samples.emplace_back(x, y);
        }
    }

    return samples;
}

void save_samples(const std::vector<Eigen::Vector2f>& samples, const std::string& path) {
    std::ofstream file(path);
    file << "x,y\n";
    file << std::fixed << std::setprecision(6);
    for (const auto& p : samples) {
        file << p.x() << "," << p.y() << "\n";
    }
}

void save_graph(const lbg::LBG2f& model, const std::string& nodes_path, const std::string& edges_path) {
    std::vector<Eigen::Vector2f> nodes;
    std::vector<std::pair<int, int>> edges;
    model.get_graph(nodes, edges);

    {
        std::ofstream file(nodes_path);
        file << "x,y\n";
        file << std::fixed << std::setprecision(6);
        for (const auto& n : nodes) {
            file << n.x() << "," << n.y() << "\n";
        }
    }

    {
        std::ofstream file(edges_path);
        file << "i,j\n";
        for (const auto& e : edges) {
            file << e.first << "," << e.second << "\n";
        }
    }
}

int main(int argc, char* argv[]) {
    // Default parameters
    std::string output_dir = "lbg_triple_ring_output";
    int n_samples = 1500;
    int n_iterations = 100;  // LBG uses epochs, not iterations
    unsigned int seed = 42;
    int frame_interval = 1;  // Save every epoch

    if (argc > 1) output_dir = argv[1];
    if (argc > 2) n_samples = std::stoi(argv[2]);
    if (argc > 3) n_iterations = std::stoi(argv[3]);
    if (argc > 4) seed = static_cast<unsigned int>(std::stoi(argv[4]));

    fs::create_directories(output_dir);

    std::cout << "LBG Triple Ring Test (C++)\n";
    std::cout << "==========================\n";
    std::cout << "Output directory: " << output_dir << "\n";
    std::cout << "Samples: " << n_samples << "\n";
    std::cout << "Epochs: " << n_iterations << "\n";
    std::cout << "Seed: " << seed << "\n\n";

    // Generate triple ring samples
    std::cout << "Generating samples...\n";
    auto samples = sample_triple_ring(n_samples, seed);
    save_samples(samples, output_dir + "/samples.csv");
    std::cout << "Generated " << samples.size() << " samples\n\n";

    // Setup LBG (parameters matching Python version)
    lbg::LBGParams params;
    params.n_nodes = 50;
    params.max_epochs = n_iterations;
    params.convergence_threshold = 1e-6f;
    params.use_utility = false;

    std::cout << "LBG Parameters:\n";
    std::cout << "  n_nodes: " << params.n_nodes << "\n";
    std::cout << "  max_epochs: " << params.max_epochs << "\n";
    std::cout << "  convergence_threshold: " << params.convergence_threshold << "\n";
    std::cout << "  use_utility: " << (params.use_utility ? "true" : "false") << "\n\n";

    lbg::LBG2f model(params, seed);
    model.init_from_data(samples);

    std::cout << "Training...\n";

    std::vector<int> frame_iterations;

    // Measure training time
    auto start_time = std::chrono::high_resolution_clock::now();

    auto callback = [&](const lbg::LBG2f& lbg, int epoch) {
        if (epoch % frame_interval == 0 || epoch == n_iterations - 1) {
            std::stringstream ss;
            ss << std::setfill('0') << std::setw(5) << epoch;
            std::string suffix = ss.str();

            save_graph(lbg, output_dir + "/nodes_" + suffix + ".csv",
                       output_dir + "/edges_" + suffix + ".csv");

            frame_iterations.push_back(epoch);

            std::cout << "Epoch " << epoch << ": "
                      << lbg.num_nodes() << " nodes, "
                      << lbg.num_edges() << " edges\n";
        }
    };

    model.train(samples, n_iterations, callback);

    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);

    // Save final result
    save_graph(model, output_dir + "/final_nodes.csv", output_dir + "/final_edges.csv");

    // Save frame list
    {
        std::ofstream file(output_dir + "/frames.csv");
        file << "iteration\n";
        for (int iter : frame_iterations) {
            file << iter << "\n";
        }
    }

    // Save metadata
    {
        std::ofstream file(output_dir + "/metadata.csv");
        file << "key,value\n";
        file << "algorithm,LBG\n";
        file << "n_epochs," << n_iterations << "\n";
        file << "time_ms," << duration.count() << "\n";
        file << "final_nodes," << model.num_nodes() << "\n";
        file << "final_edges," << model.num_edges() << "\n";
    }

    std::cout << "\nTraining complete!\n";
    std::cout << "Time: " << duration.count() << " ms\n";
    std::cout << "Final: " << model.num_nodes() << " nodes, " << model.num_edges() << " edges\n";
    std::cout << "Results saved to: " << output_dir << "\n";

    return 0;
}
